base/parse.jl

```diff
@@ -218,6 +218,93 @@ function parse(::Type{T}, s::AbstractString; base::Union{Nothing,Integer} = noth
 end
 
 ## string to float functions ##
+function parsefloat_preamble(s::AbstractString, base::Int, i::Int=1)
+    isempty(s) && throw(ArgumentError("Empty string!"))
+    c = ' '
+    sign = 1
+    while isspace(c)
+        c, i = next(s,i)
```

Comment:
This is abusing the iteration protocol.  It is assuming that `start(s)` is the same as the starting index `i=1`.

Better to use `i = nextind(s,i); c = @inbounds s[i]`

Reply:
See also #25833.

You should also initialize `i` to `firstindex(s)`.

Reply:
This is also true for `parseint_preamble` where i drew inspiration from.

Should i update only the code from this PR to not abuse the iteration protocol?
I could also just leave it like this and make a seperate PR to update the whole `parse.jl` file.

Reply:
@Keno, do you have a preference on how to proceed here?

Reply:
I replaced `next()` and `done()` calls in the whole file - this shouldn't 'abuse' the iteration protocol any longer.